File: ruby/primes.rb
def prime?(n)
  return false if 1 == n
  # test 2 first to start with 3
  return true  if 2 == n
  return false if n.even?
  # start with 3 to skip evens
  divisor = 3
  while (divisor * divisor) <= n
    return false if n % divisor == 0
    # skip evens to go twice as fast
    divisor += 2
  end
  true
end

def primes(from: 2, to: Float::INFINITY)
  enumerator = if to > from then from.upto(to) else from.downto(to) end
  enumerator.lazy.select { |n| prime? n }
end

File: erlang/primes.erl
% -module(primes).
% -export([is_prime/1]).

is_prime(X) when X =< 0 ->
  throw("Number must be a positive integer.");
is_prime(1) ->
  false;
% test for 2 to start the loop with 3
is_prime(2) ->
  true;
is_prime(X) when X rem 2 == 0 ->
  false;
% start with 3 to be able to skip evens
is_prime(X) when X >= 3 ->
  is_prime(X, 3).

is_prime(X, Factor) when (Factor * Factor) > X ->
  true;
is_prime(X, Factor) when X rem Factor == 0 ->
  false;
% skip evens to go twice as fast
is_prime(X, Factor) ->
  is_prime(X, Factor + 2).
